src/Network/Anonymous/Tor/Protocol.hs

```diff
@@ -245,3 +246,17 @@ mapOnion s rport lport = do
   reply <- sendCommand s (BS8.concat ["ADD_ONION NEW:BEST Port=", BS8.pack (show rport), ",127.0.0.1:", BS8.pack(show lport), "\n"])
 
   return . B32.b32String' . fromJust . Ast.tokenValue . head . Ast.lineMessage . fromJust $ Ast.line (BS8.pack "ServiceID") reply
+
+-- | Creates a hidden service from an existing private key in base64 and maps
+--   a public port to a local port. Useful for bridging a local service (e.g. a
+--   webserver or irc daemon) as a Tor hidden service.
+mapPersistentOnion :: MonadIO m
+         => Network.Socket     -- ^ Connection with tor Control port
+         -> Integer            -- ^ Remote point of hidden service to listen at
+         -> Integer            -- ^ Local port to map onion service to
+         -> BS.ByteString      -- ^ Private key in Base64
+         -> m B32.Base32String -- ^ The address/service id of the Onion without the .onion aprt
+mapPersistentOnion s rport lport pkey = do
+  reply <- sendCommand s (BS8.pack "ADD_ONION RSA1024:" `BS.append` pkey `BS.append` BS8.concat [" Port=", BS8.pack (show rport), ",127.0.0.1:", BS8.pack(show lport), "\n"])
```

Comment:
Why do you first use `append` and then `concat`? Couldn’t the whole expression inside the brackets be a `concat`? Another option would be `Data.Monoid.(<>)`.